### ISSUE TITLE
[HOLD for payment 2023-05-05] [$1000] Invalid urls in chat are parsed as valid, redirecting to the app again.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17591]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open any chat
2. Enter an invalid url with hyphen in the end(before TLD)
3. Now open that url

## Expected Results: 
should not be parsed as url, and if parsed, then should go to that invalid url ending up with an error from search engine on screen
## Actual results: 
parsed as url, and then on opening that, it redirects to the app again

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop
 
**Version Number:** 1.3.1
**Reproducible in staging?:**  y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/232833520-918be1fc-8e1b-4892-ac27-6f8db5a2c3ba.mp4

https://user-images.githubusercontent.com/43996225/232833547-93052a31-e70d-457b-9d5d-e877f643671a.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @chiragxarora
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681656328792209

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019d3fae18911c3787</li>
        <li>Upwork Job ID: 1648700378826989568</li>
        <li>Last Price Increase: 2023-04-19</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by kadiealexander** [comment]
Reassigning as I'm heading OOO for a week.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @slafortune (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by slafortune** [comment]
Looks good!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~019d3fae18911c3787

**Comment by MelvinBot** [comment]
Current assignee @slafortune is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @mananjadhav (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jasperhuangg (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by jjcoffee** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Invalid URLs are still parsed by the autolinker as if they are URLs, which results in the weird redirecting behaviour.

### What is the root cause of that problem?
The autolinker is not validating URLs in the same way as the BE, specifically it is accepting URLs that start or end with a `-` as valid. You can see if you look closely that the link initially shows as underlined and then the underline goes away (once the BE request finishes and updates the FE). Basically the BE is stripping the URL from the `href`, leaving it empty (which then causes the odd redirect behaviour).

### What changes do you think we should make in order to solve the problem?
We need to correct the autolinker's regex so that URLs that start or end with a `-` are not autolinked, as they are invalid.

We need to update the `URL_WEBSITE_REGEX` here:

https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/Url.js#L3

```
(https?:\\/\\/)?((?:www\\.)?[a-z0-9][-a-z0-9]+?[a-z0-9]\\.)+(?:${TLD_REGEX})(?:\\:\\d{2,4}|\\b|(?=_))
```

This results in `test-.com` being displayed as normal text, whereas for `-test.com`, only the valid domain part is rendered as a link (this is consistent with how we deal with `~test.com`, for example).

### What alternative solutions did you explore? (Optional)
It might be useful for a retrospective fix here too so that previously sent links that got caught up in this don't render incorrectly. This could be done either in the BE, or we can choose to not render empty links on the FE (I think this is a bit out of scope for this bug, though!).

**Comment by jasperhuangg** [comment]
@jjcoffee The URL `abc-.com` is technically still a valid URL, so I think we still want it to be parsed by the regex. If clicked we just want it to perform like a regular URL and open in a different tab.

Your solution prevents the URL from being highlighted, which isn't what we want.

**Comment by jjcoffee** [comment]
@jasperhuangg The `abc-` label in the `abc-.com` URL represents an invalid domain name (they [cannot](https://en.wikipedia.org/wiki/Domain_Name_System#cite_ref-rfc3696_28-0:~:text=Labels%20may%20not%20start%20or%20end%20with%20a%20hyphen) start or end with a hyphen), which is why (I think!) the BE is stripping it out as it's impossible for the URL to go anywhere.

If you still think we want to accept it as a valid URL, I think BE would need to be changed.

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Invalid urls in chat redirecting to the app again.
### What is the root cause of that problem?
After network debuging, it is a backend issue
```javascript
// in `api?command=AddComment` api we send correct <a> with href attribute.
reportComment: <a href="https://abc-.com" target="_blank" rel="noreferrer noopener">https://abc-.com</a>

// but in `api?command=OpenReport` is received without href attribute
`"html": "<a target="_blank" rel="noreferrer noopener">https://abc-.com</a>"`
```

### What changes do you think we should make in order to solve the problem?
The fix should be done from backend side.

### What alternative solutions did you explore? (Optional)

**Comment by MelvinBot** [comment]
@mananjadhav, @slafortune, @jasperhuangg Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by jasperhuangg** [comment]
ah @jjcoffee I see what you mean, thanks I didn't know! If that's the case I'm happy to move forward with your solution.

**Comment by MelvinBot** [comment]
📣 @jjcoffee You have been assigned to this job by @jasperhuangg!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by chiragxarora** [comment]
Hey @jasperhuangg , 
did you see @ahmedGaber93 's proposal?

**Comment by jasperhuangg** [comment]
I did, but not sure how that solves the problem. Wouldn't adding an `href` tag make it a clickable link, which isn't what we want to happen? @chiragxarora 

It seems the back-end already doesn't recognize these as links, and we need to update our front-end to do so as well.

**Comment by chiragxarora** [comment]
I think this can be solved in 2 ways. Either we don't parse it as url on frontend, or we let it be and let user go to abc-.com which shows an error from search engine. Main problem is that the link is redirecting us to app. We can also consider the option of letting it open in a new tab too. That's why I mentioned his proposal

**Comment by jjcoffee** [comment]
@jasperhuangg No worries, I had to look it up myself! 

The [PR](https://github.com/Expensify/expensify-common/pull/523) is now ready for review. cc @mananjadhav

**Comment by jjcoffee** [comment]
@jasperhuangg @mananjadhav The [PR](https://github.com/Expensify/App/pull/18020) to update the expensify-common package in the app is ready for review.

**Comment by jjcoffee** [comment]
PR against the app closed out as another PR already got merged with a newer version of expensify-common. Will update when that PR gets deployed to production.

**Comment by jjcoffee** [comment]
@slafortune Are you able to hire me on the [Upwork job](https://www.upwork.com/jobs/~019d3fae18911c3787) ready for future payment, when you get a chance?

**Comment by chiragxarora** [comment]
I also haven't been hired for the bug report yet and sadly I don't have much connects to apply. Could you please send me an offer instead?

My profile link: https://www.upwork.com/freelancers/~014a1b566a7d8ee0c5?s=1110580755107926016

**Comment by mananjadhav** [comment]
@jjcoffee I thought we're covering both the cases. Leading and trailing hyphens. It looks like leading hyphen is still being rendered as a link.

<img width="1036" alt="image" src="https://user-images.githubusercontent.com/3069065/234961797-069e3691-1b87-485b-8069-5f21c4d9991c.png">

**Comment by jjcoffee** [comment]
@mananjadhav This is expected (only the URL part is highlighted, without the hyphen), as stated in my [proposal](https://github.com/Expensify/App/issues/17591#issuecomment-1515032004):

> This results in test-.com being displayed as normal text, whereas for -test.com, only the valid domain part is rendered as a link (this is consistent with how we deal with ~test.com, for example).

**Comment by jjcoffee** [comment]
The other PR with the newer version of expensify-common that also has my fix has now been [deployed](https://github.com/Expensify/App/pull/17564#issuecomment-1527141563) to production. If no regressions arise, payment should be issued on 2023-05-05. cc @slafortune

**Comment by chiragxarora** [comment]
Hi @slafortune could you pls update this issue?

**Comment by mananjadhav** [comment]
@mountiny @jasperhuangg I unable to pinpoint the offending PR as it looks like it didn't work since a long time and not due to the recent changes.

**Here's the regression test proposal.**

1. Open any chat
2. Enter an invalid URL ending with a hyphen (e.g., abc-.com) in the message and send
3. Verify that the message is not hyperlinked
4. Ensure the message is successfully sent to the API and then still the message shouldn't be hyperlink
5. Now enter another invalid URL starting with a hyphen e.g -abc.com in the message and send
6. Verify that excluding the hyphen, the rest of the text is highlighted as hyperlink
7. Again wait for the message to be sent via API and verify the behavior as point 6.


Do we agree 👍  or 👎 ?

What do you folks think @jasperhuangg @mountiny @jjcoffee @slafortune ?

**Comment by jjcoffee** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@**[mananjadhav](https://github.com/mananjadhav)**] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@**[mananjadhav](https://github.com/mananjadhav)**] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@**[mananjadhav](https://github.com/mananjadhav)**] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@**[mananjadhav](https://github.com/mananjadhav)**] Determine if we should create a regression test for this bug.
- [ ] [@**[mananjadhav](https://github.com/mananjadhav)**] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [**[@slafortune](https://github.com/slafortune)**] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by jjcoffee** [comment]
50% timeliness bonus is due as this was assigned 24 April, PR (in expensify-common) merged April 25. So I make it $1,500 for contributors (@jjcoffee @mananjadhav) and $250 for the reporting bonus to @chiragxarora. I have applied to the Upwork job @slafortune.

**Comment by jjcoffee** [comment]
@chiragxarora I meant @mananjadhav is the C+

**Comment by slafortune** [comment]
@mananjadhav - can you check off any pertaining steps in the BZ checklist above?
@chiragxarora @mananjadhav @jjcoffee contracts are sent 👍

**Comment by jjcoffee** [comment]
@slafortune Thanks Stevie - contract accepted!

**Comment by mananjadhav** [comment]
@slafortune Can you please refer this [comment](https://github.com/Expensify/App/issues/17591#issuecomment-1533192902) for the checklist? Parsing invalid URL seems to be existing since a long time.

**Comment by slafortune** [comment]
@mananjadhav Yep, I agree with the regression test and that's been created.  Is there a PR that introduced the bug has been identified?

**Comment by mananjadhav** [comment]
I couldn't trace an offending PR. Plus as I mentioned it looks like we've had the issue since inception.
